<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GBModels.jl</title><meta name="title" content="Home · GBModels.jl"/><meta property="og:title" content="Home · GBModels.jl"/><meta property="twitter:title" content="Home · GBModels.jl"/><meta name="description" content="Documentation for GBModels.jl."/><meta property="og:description" content="Documentation for GBModels.jl."/><meta property="twitter:description" content="Documentation for GBModels.jl."/><meta property="og:url" content="https://GenomicBreeding.github.io/GBModels.jl/"/><meta property="twitter:url" content="https://GenomicBreeding.github.io/GBModels.jl/"/><link rel="canonical" href="https://GenomicBreeding.github.io/GBModels.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GBModels.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GBModels.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GBModels.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GBModels"><a class="docs-heading-anchor" href="#GBModels">GBModels</a><a id="GBModels-1"></a><a class="docs-heading-anchor-permalink" href="#GBModels" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/GenomicBreeding/GBModels.jl">GBModels</a>.</p><ul><li><a href="#GBModels.LπDist"><code>GBModels.LπDist</code></a></li><li><a href="#GBModels.NπDist"><code>GBModels.NπDist</code></a></li><li><a href="#GBModels.TπDist"><code>GBModels.TπDist</code></a></li><li><a href="#Base.maximum-Tuple{GBModels.LπDist}"><code>Base.maximum</code></a></li><li><a href="#Base.maximum-Tuple{GBModels.TπDist}"><code>Base.maximum</code></a></li><li><a href="#Base.maximum-Tuple{GBModels.NπDist}"><code>Base.maximum</code></a></li><li><a href="#Base.minimum-Tuple{GBModels.TπDist}"><code>Base.minimum</code></a></li><li><a href="#Base.minimum-Tuple{GBModels.NπDist}"><code>Base.minimum</code></a></li><li><a href="#Base.minimum-Tuple{GBModels.LπDist}"><code>Base.minimum</code></a></li><li><a href="#Base.rand-Tuple{Random.AbstractRNG, GBModels.NπDist}"><code>Base.rand</code></a></li><li><a href="#Base.rand-Tuple{Random.AbstractRNG, GBModels.LπDist}"><code>Base.rand</code></a></li><li><a href="#Base.rand-Tuple{Random.AbstractRNG, GBModels.TπDist}"><code>Base.rand</code></a></li><li><a href="#Distributions.logpdf-Tuple{GBModels.NπDist, Real}"><code>Distributions.logpdf</code></a></li><li><a href="#Distributions.logpdf-Tuple{GBModels.LπDist, Real}"><code>Distributions.logpdf</code></a></li><li><a href="#Distributions.logpdf-Tuple{GBModels.TπDist, Real}"><code>Distributions.logpdf</code></a></li><li><a href="#GBModels.bayesian-Tuple{}"><code>GBModels.bayesian</code></a></li><li><a href="#GBModels.lasso-Tuple{}"><code>GBModels.lasso</code></a></li><li><a href="#GBModels.ols-Tuple{}"><code>GBModels.ols</code></a></li><li><a href="#GBModels.ridge-Tuple{}"><code>GBModels.ridge</code></a></li><li><a href="#GBModels.turing_bayesG-Tuple{Any, Any}"><code>GBModels.turing_bayesG</code></a></li><li><a href="#GBModels.turing_bayesGs-Tuple{Any, Any}"><code>GBModels.turing_bayesGs</code></a></li><li><a href="#GBModels.turing_bayesL-Tuple{Any, Any}"><code>GBModels.turing_bayesL</code></a></li><li><a href="#GBModels.turing_bayesLs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T"><code>GBModels.turing_bayesLs</code></a></li><li><a href="#GBModels.turing_bayesLπ-Tuple{Any, Any}"><code>GBModels.turing_bayesLπ</code></a></li><li><a href="#GBModels.turing_bayesLπs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T"><code>GBModels.turing_bayesLπs</code></a></li><li><a href="#GBModels.turing_bayesNπ-Tuple{Any, Any}"><code>GBModels.turing_bayesNπ</code></a></li><li><a href="#GBModels.turing_bayesNπs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T"><code>GBModels.turing_bayesNπs</code></a></li><li><a href="#GBModels.turing_bayesT-Tuple{Any, Any}"><code>GBModels.turing_bayesT</code></a></li><li><a href="#GBModels.turing_bayesTπ-Tuple{Any, Any}"><code>GBModels.turing_bayesTπ</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.LπDist" href="#GBModels.LπDist"><code>GBModels.LπDist</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Laplace distribution with a point mass at 0.0</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L490-L492">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.NπDist" href="#GBModels.NπDist"><code>GBModels.NπDist</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Gaussian distribution with a point mass at 0.0</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L341-L343">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.TπDist" href="#GBModels.TπDist"><code>GBModels.TπDist</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>T-distribution with a point mass at 0.0</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L239-L241">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.maximum-Tuple{GBModels.LπDist}" href="#Base.maximum-Tuple{GBModels.LπDist}"><code>Base.maximum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Maximum value of the LπDist distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L542-L544">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.maximum-Tuple{GBModels.NπDist}" href="#Base.maximum-Tuple{GBModels.NπDist}"><code>Base.maximum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Maximum value of the NπDist distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L393-L395">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.maximum-Tuple{GBModels.TπDist}" href="#Base.maximum-Tuple{GBModels.TπDist}"><code>Base.maximum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Maximum value of the TπDist distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L290-L292">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.minimum-Tuple{GBModels.LπDist}" href="#Base.minimum-Tuple{GBModels.LπDist}"><code>Base.minimum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Minimum value of the LπDist distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L537-L539">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.minimum-Tuple{GBModels.NπDist}" href="#Base.minimum-Tuple{GBModels.NπDist}"><code>Base.minimum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Minimum value of the NπDist distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L388-L390">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.minimum-Tuple{GBModels.TπDist}" href="#Base.minimum-Tuple{GBModels.TπDist}"><code>Base.minimum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Minimum value of the TπDist distribution</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L285-L287">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.rand-Tuple{Random.AbstractRNG, GBModels.LπDist}" href="#Base.rand-Tuple{Random.AbstractRNG, GBModels.LπDist}"><code>Base.rand</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Sampling method for LπDist</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">d = LπDist(0.1, 0.0, 1.0)
rand(d)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L499-L507">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.rand-Tuple{Random.AbstractRNG, GBModels.NπDist}" href="#Base.rand-Tuple{Random.AbstractRNG, GBModels.NπDist}"><code>Base.rand</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Sampling method for NπDist</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">d = NπDist(0.1, 0.0, 1.0)
rand(d)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L350-L358">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.rand-Tuple{Random.AbstractRNG, GBModels.TπDist}" href="#Base.rand-Tuple{Random.AbstractRNG, GBModels.TπDist}"><code>Base.rand</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Sampling method for TπDist</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">d = TπDist(0.1, 1.0)
rand(d)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L247-L255">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Distributions.logpdf-Tuple{GBModels.LπDist, Real}" href="#Distributions.logpdf-Tuple{GBModels.LπDist, Real}"><code>Distributions.logpdf</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>log(pdf) of LπDist</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">d = LπDist(0.1, 0.0, 1.0)
logpdf.(d, [-1.0, 0.0, 1.0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L518-L526">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Distributions.logpdf-Tuple{GBModels.NπDist, Real}" href="#Distributions.logpdf-Tuple{GBModels.NπDist, Real}"><code>Distributions.logpdf</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>log(pdf) of NπDist</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">d = NπDist(0.1, 0.0, 1.0)
logpdf.(d, [-1.0, 0.0, 1.0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L369-L377">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Distributions.logpdf-Tuple{GBModels.TπDist, Real}" href="#Distributions.logpdf-Tuple{GBModels.TπDist, Real}"><code>Distributions.logpdf</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>log(pdf) of TπDist</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">d = TπDist(0.1, 1.0)
logpdf.(d, [-1.0, 0.0, 1.0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L266-L274">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.bayesian-Tuple{}" href="#GBModels.bayesian-Tuple{}"><code>GBModels.bayesian</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">bayesian(;
    genomes::Genomes,
    phenomes::Phenomes,
    trait_idx::Int64 = 1,
    turing_model::Function=turing_bayesG,
    seed::Int64 = 123,
    nburnin::Int64 = 500,
    niter::Int64 = 1_500,
    verbose::Bool = false,
)::Fit</code></pre><p>Fit a Bayesian ridge regression model</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, l=1_000, verbose=false);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.1 0.01 0.01;], verbose=false);

julia&gt; tebv = GBCore.analyse(trials, max_levels=5, max_time_per_model=2, verbose=false);

julia&gt; phenomes = tebv.phenomes[1];

julia&gt; fit = Suppressor.@suppress bayesian(genomes=genomes, phenomes=phenomes);

julia&gt; fit.metrics[&quot;cor&quot;] &gt; 0.5
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/linear.jl#L247-L276">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.lasso-Tuple{}" href="#GBModels.lasso-Tuple{}"><code>GBModels.lasso</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">lasso(genomes::Genomes, phenomes::Phenomes, trait_idx::Int64=1, verbose::Bool=false)::Fit</code></pre><p>Fit a LASSO (least absolute shrinkage and selection operator; L1) regression model</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, l=100, verbose=false);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.1 0.01 0.01;], verbose=false);

julia&gt; phenomes = Phenomes(n=10, t=1);

julia&gt; phenomes.entries = trials.entries[1:10]; phenomes.populations = trials.populations[1:10]; phenomes.traits = trials.traits; phenomes.phenotypes = trials.phenotypes[1:10, :];

julia&gt; fit = lasso(genomes=genomes, phenomes=phenomes);

julia&gt; fit.metrics[&quot;cor&quot;] &gt; 0.0
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/linear.jl#L156-L176">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.ols-Tuple{}" href="#GBModels.ols-Tuple{}"><code>GBModels.ols</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ols(genomes::Genomes, phenomes::Phenomes, trait_idx::Int64=1, verbose::Bool=false)::Fit</code></pre><p>Fit an ordinary least squares model</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, l=100, verbose=false);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.1 0.01 0.01;], verbose=false);

julia&gt; phenomes = Phenomes(n=10, t=1);

julia&gt; phenomes.entries = trials.entries[1:10]; phenomes.populations = trials.populations[1:10]; phenomes.traits = trials.traits; phenomes.phenotypes = trials.phenotypes[1:10, :];

julia&gt; fit = ols(genomes=genomes, phenomes=phenomes);

julia&gt; fit.model == &quot;ols&quot;
true

julia&gt; fit.metrics[&quot;cor&quot;] &gt; 0.50
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/linear.jl#L1-L24">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.ridge-Tuple{}" href="#GBModels.ridge-Tuple{}"><code>GBModels.ridge</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ridge(genomes::Genomes, phenomes::Phenomes, trait_idx::Int64=1, verbose::Bool=false)::Fit</code></pre><p>Fit a ridge (L2) regression model</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = GBCore.simulategenomes(n=10, l=100, verbose=false);

julia&gt; trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.1 0.01 0.01;], verbose=false);

julia&gt; phenomes = Phenomes(n=10, t=1);

julia&gt; phenomes.entries = trials.entries[1:10]; phenomes.populations = trials.populations[1:10]; phenomes.traits = trials.traits; phenomes.phenotypes = trials.phenotypes[1:10, :];

julia&gt; fit = ridge(genomes=genomes, phenomes=phenomes);

julia&gt; fit.metrics[&quot;cor&quot;] &gt; 0.50
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/linear.jl#L71-L91">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesG-Tuple{Any, Any}" href="#GBModels.turing_bayesG-Tuple{Any, Any}"><code>GBModels.turing_bayesG</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Gaussian prior with common variance</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Benchmarking
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
model = turing_bayesG(G, y)
benchmarks = TuringBenchmarking.benchmark_model(
    model;
    # Check correctness of computations
    check=true,
    # Automatic differentiation backends to check and benchmark
    adbackends=[:forwarddiff, :reversediff, :reversediff_compiled, :zygote]
)


# Test more loci
genomes = GBCore.simulategenomes(n=10, l=10_000)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Check for uninferred types in the model
@code_warntype model = turing_bayesG(G, y)
# Fit
model = turing_bayesG(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L1-L47">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesGs-Tuple{Any, Any}" href="#GBModels.turing_bayesGs-Tuple{Any, Any}"><code>GBModels.turing_bayesGs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Gaussian prior with varying variances</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesGs(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L62-L90">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesL-Tuple{Any, Any}" href="#GBModels.turing_bayesL-Tuple{Any, Any}"><code>GBModels.turing_bayesL</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Laplacian prior with a common scale</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesL(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L107-L135">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesLs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T" href="#GBModels.turing_bayesLs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T"><code>GBModels.turing_bayesLs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Laplacian prior with varying scales</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesLs(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L150-L178">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesLπ-Tuple{Any, Any}" href="#GBModels.turing_bayesLπ-Tuple{Any, Any}"><code>GBModels.turing_bayesLπ</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Laplacian prior with a point mass at zero and common scale</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesLπ(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L548-L576">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesLπs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T" href="#GBModels.turing_bayesLπs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T"><code>GBModels.turing_bayesLπs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Laplacian prior with a point mass at zero and common scale</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesLπs(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L591-L619">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesNπ-Tuple{Any, Any}" href="#GBModels.turing_bayesNπ-Tuple{Any, Any}"><code>GBModels.turing_bayesNπ</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Gaussian prior with a point mass at zero and common variance</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesNπ(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L399-L427">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesNπs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T" href="#GBModels.turing_bayesNπs-Union{Tuple{T}, Tuple{Any, Any}, Tuple{Any, Any, DynamicPPL.TypeWrap{T}}} where T"><code>GBModels.turing_bayesNπs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a Gaussian prior with a point mass at zero and varying variances</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesNπs(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L442-L470">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesT-Tuple{Any, Any}" href="#GBModels.turing_bayesT-Tuple{Any, Any}"><code>GBModels.turing_bayesT</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a T-distribution</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesT(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L197-L225">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBModels.turing_bayesTπ-Tuple{Any, Any}" href="#GBModels.turing_bayesTπ-Tuple{Any, Any}"><code>GBModels.turing_bayesTπ</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Turing specification of Bayesian linear regression using a T-distribution with a point mass at zero</p><p><strong>Example usage</strong></p><pre><code class="language-julia hljs"># Simulate data
genomes = GBCore.simulategenomes(n=10, l=100)
trials, _ = GBCore.simulatetrials(genomes=genomes, n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=3, f_add_dom_epi=[0.9 0.01 0.00;])
tebv = GBCore.analyse(trials, max_levels = 15, max_time_per_model = 2)
phenomes = tebv.phenomes[1]
# Extract genotype and phenotype data
G::Matrix{Float64} = genomes.allele_frequencies
y::Vector{Float64} = phenomes.phenotypes[:, 1]
# Regress for just 200 iterations for demonstration purposes only. Use way way more iterations, e.g. 10,000.
model = turing_bayesTπ(G, y)
# We use compile=true in AutoReverseDiff() because we do not have any if-statements in our Turing model below
rng::TaskLocalRNG = Random.seed!(123)
niter::Int64 = 1_500
nburnin::Int64 = 500
@time chain = Turing.sample(rng, model, NUTS(nburnin, 0.5, max_depth=5, Δ_max=1000.0, init_ϵ=0.2; adtype=AutoReverseDiff(compile=true)), niter-nburnin, progress=true);
# Use the mean paramter values after 150 burn-in iterations
params = Turing.get_params(chain[150:end, :, :]);
b_hat = vcat(mean(params.intercept), mean(stack(params.coefficients, dims=1)[:, :, 1], dims=2)[:,1]);
# Assess prediction accuracy
y_pred::Vector{Float64} = hcat(ones(size(G,1)), G) * b_hat;
UnicodePlots.scatterplot(y, y_pred)
performance::Dict{String, Float64} = metrics(y, y_pred)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBModels.jl/blob/c14f34e3dbeb29f0610cb6d174fd7796c45cbc90/src/bayes.jl#L296-L324">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Saturday 11 January 2025 04:05">Saturday 11 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
